File: lib/core/utils/url_extractor.dart
import '../constants/threat_patterns.dart';

class UrlExtractor {
  UrlExtractor._();

  static final _urlRegex = RegExp(
    ThreatPatterns.urlPattern,
    caseSensitive: false,
    multiLine: true,
  );

  /// Extract all URLs from a text string
  static List<String> extractUrls(String text) {
    if (text.isEmpty) return [];
    final matches = _urlRegex.allMatches(text);
    return matches
        .map((m) => m.group(0)!.trim())
        .where((url) => url.length > 4)
        .toList();
  }

  /// Get domain from URL
  static String? getDomain(String url) {
    try {
      String cleaned = url;
      if (!cleaned.startsWith('http')) {
        cleaned = 'https://$cleaned';
      }
      final uri = Uri.tryParse(cleaned);
      if (uri == null) return null;
      String host = uri.host.toLowerCase();
      if (host.startsWith('www.')) host = host.substring(4);
      return host;
    } catch (_) {
      return null;
    }
  }

  /// Get TLD from domain
  static String? getTld(String domain) {
    final parts = domain.split('.');
    if (parts.length < 2) return null;
    return '.${parts.last}';
  }

  /// Normalize URL for analysis
  static String normalize(String url) {
    String u = url.trim().toLowerCase();
    if (!u.startsWith('http')) u = 'https://$u';
    return u;
  }

  /// Whether the URL hides structure behind encoding or a second scheme.
  ///
  /// Percent-encoding is matched case-insensitively. It was written as `%2F` /
  /// `%3A`, but the only caller — `PhishingRepository.analyzeUrl` — runs
  /// [normalize] first, which lowercases, so those two branches could never
  /// match and the rule silently scored nothing for the life of the app. RFC
  /// 3986 permits either case in an escape sequence, so a fixed-case check was
  /// never right regardless of the lowercasing.
  ///
  /// The `//` clause looks for a *second* occurrence, past the scheme's own —
  /// `https://a.com/?next=https://evil.tk` — which is the redirect-wrapper
  /// shape this is meant to catch.
  static bool hasEncodedTricks(String url) {
    final lower = url.toLowerCase();
    return lower.contains('%2f') ||
        lower.contains('%3a') ||
        url.contains('@') ||
        url.contains('//') && url.indexOf('//') != url.lastIndexOf('//');
  }

  /// Check if URL uses IP address instead of domain
  static bool isIpAddress(String url) {
    final ipPattern = RegExp(r'https?://\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}');
    return ipPattern.hasMatch(url);
  }

  /// Count subdomains (many subdomains can indicate phishing)
  static int subdomainCount(String domain) {
    return domain.split('.').length - 2;
  }

  /// Check if URL is very long (phishing indicator)
  static bool isExcessivelyLong(String url) {
    return url.length > 100;
  }
}

File: web/server/src/engines/urlExtractor.js
import { urlPattern } from './threatPatterns.js';

/** Port of `lib/core/utils/url_extractor.dart`. */

const urlRegex = new RegExp(urlPattern, 'gim');

export function extractUrls(text) {
  if (!text) return [];
  // A /g regex carries lastIndex between calls, so it is rebuilt per call
  // rather than shared — otherwise the second scan of the same text silently
  // starts partway through and misses the first link.
  return [...text.matchAll(new RegExp(urlRegex.source, 'gim'))]
    .map((m) => m[0].trim())
    .filter((url) => url.length > 4);
}

export function getDomain(url) {
  try {
    // The scheme test is case-insensitive here where the Dart original is not.
    // In the scan pipeline it makes no difference — `analyzeUrl` runs
    // `normalize`, which lowercases, before this is ever called — but
    // `getDomain` is also exported on its own, and given `HTTPS://example.com`
    // the case-sensitive form prepends a second scheme and returns "https" as
    // the domain. Defensive, not divergent.
    const cleaned = /^https?:\/\//i.test(url) ? url : `https://${url}`;
    const parsed = new URL(cleaned);
    let host = parsed.hostname.toLowerCase();
    if (host.startsWith('www.')) host = host.slice(4);
    return host;
  } catch {
    return null;
  }
}

export function getTld(domain) {
  const parts = domain.split('.');
  if (parts.length < 2) return null;
  return `.${parts[parts.length - 1]}`;
}

export function normalize(url) {
  const u = url.trim().toLowerCase();
  return u.startsWith('http') ? u : `https://${u}`;
}

/**
 * Whether the URL hides structure behind encoding or a second scheme.
 *
 * Percent-encoding is matched case-insensitively — see the Dart original for
 * why the fixed-case form never fired. RFC 3986 permits either case in an
 * escape sequence.
 */
export function hasEncodedTricks(url) {
  const lower = url.toLowerCase();
  return (
    lower.includes('%2f') ||
    lower.includes('%3a') ||
    url.includes('@') ||
    (url.includes('//') && url.indexOf('//') !== url.lastIndexOf('//'))
  );
}

export function isIpAddress(url) {
  return /https?:\/\/\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}/.test(url);
}

export function subdomainCount(domain) {
  return domain.split('.').length - 2;
}

export function isExcessivelyLong(url) {
  return url.length > 100;
}
